// tango/tango.db.mongo.js
var dbBase = require('./tango.db.base');
var mongodb = require('mongodb');
var BSON = mongodb.BSONPure;
var server;
var db;

mongo = function() {	
}
util.inherits(mongo, dbBase);


mongo.prototype.init = function(host, port) {
	server = new mongodb.Server(host, port, {auto_reconnect: true});
}

mongo.prototype.connect = function(database, username, password) {
	db = new mongodb.Db(database, server, {safe: true});
	db.open(function(err, db) {
		if(!err) {
			console.log('INFO: Connected to database "' + database + '"');
		} else {
			throw err;
		}
	});
}

mongo.prototype.find = function(model, filer, callback) {
	var collection = db.collection(model);
	collection.find().toArray(callback); 
}

mongo.prototype.findById = function(model, id, callback) {
	var collection = db.collection(model);
	collection.findOne({'_id':new BSON.ObjectID(id)}, callback);
}

mongo.prototype.insert = function(model, items, callback) {
	var collection = db.collection(model);
	collection.insert(items, {safe:true}, callback);
}

mongo.prototype.remove = function(model, filer, callback) {
	var collection = db.collection(model);
	collection.remove(filer, {safe:true}, callback);
}

mongo.prototype.removeById = function(model, id, callback) {
	var collection = db.collection(model); 
	collection.remove({'_id':new BSON.ObjectID(id)}, {safe:true}, callback);
}

mongo.prototype.updateById = function(model, id, item, callback) {
	var collection = db.collection(model);
	collection.update({'_id':new BSON.ObjectID(id)}, item, {safe:true}, callback);
}

module.exports = mongo;
